modules/nrfxlib/nrf_802154/sl/platform/nrf_802154_platform_sl_lptimer_zephyr.c

```diff
@@ -201,3 +256,101 @@ uint64_t nrf_802154_platform_sl_lptimer_sync_lpticks_get(void)
 {
 	return m_sync_timer.target_time;
 }
+
+nrf_802154_sl_lptimer_platform_result_t nrf_802154_platform_sl_lptimer_hw_task_prepare(
+	uint64_t fire_lpticks,
+	uint32_t ppi_channel)
+{
+	bool done_on_time;
+	nrf_802154_sl_mcu_critical_state_t mcu_cs_state;
+
+	if (!hw_task_state_set(HW_TASK_STATE_IDLE, HW_TASK_STATE_SETTING_UP)) {
+		/* The only one available set of peripherals is already used. */
+		return NRF_802154_SL_LPTIMER_PLATFORM_NO_RESOURCES;
+	}
+
+	/* Allocate the channel if not already done. */
+	if (m_hw_task.chan == RTC_CHAN_INVALID) {
+		/* The channel allocation cannot take place during module
+		 * initialization, because for nRF53 it would run
+		 * out of resources - some other module temporarily needs
+		 * an rtc channel to initialize. For this reason, this is where
+		 * the allocation is made.
+		 * Once a channel has been successfully allocated, it will be held
+		 * until nrf_802154_platform_sl_lp_timer_deinit is called.
+		 */
+		m_hw_task.chan = z_nrf_rtc_timer_chan_alloc();
+		if (m_hw_task.chan < 0) {
+			m_hw_task.chan = RTC_CHAN_INVALID;
+			hw_task_state_set(HW_TASK_STATE_SETTING_UP, HW_TASK_STATE_IDLE);
+			return NRF_802154_SL_LPTIMER_PLATFORM_NO_RESOURCES;
+		}
+	}
+
+	if (z_nrf_rtc_timer_set(m_hw_task.chan, fire_lpticks, NULL, NULL) != 0) {
+		hw_task_state_set(HW_TASK_STATE_SETTING_UP, HW_TASK_STATE_IDLE);
+		return NRF_802154_SL_LPTIMER_PLATFORM_TOO_DISTANT;
+	}
+
+	if (ppi_channel != NRF_802154_SL_HW_TASK_PPI_INVALID) {
+		cc_bind_to_ppi(m_hw_task.chan, ppi_channel);
+	}
+	m_hw_task.ppi = ppi_channel;
+
+	nrf_802154_sl_mcu_critical_enter(mcu_cs_state);
+
+	/* For triggering to take place a safe margin is 2 lpticks from `now`. */
+	if ((z_nrf_rtc_timer_read() + 2) > fire_lpticks) {
+		/* it is too late */
+		cc_unbind(m_hw_task.chan, ppi_channel);
+		m_hw_task.ppi = NRF_802154_SL_HW_TASK_PPI_INVALID;
+		z_nrf_rtc_timer_abort(m_hw_task.chan);
```

Comment:
I'm not a fan of calling external modules from critical sections. Would it be OK to exit the critical section before calling `cc_unbind()` above?

Reply:
sure, i'll move that beyond the critical section

Reply:
After all, it turned out that the critical section was not needed at all :)
Removed

Reply:
I think there is a race condition here. It is possible that PPI was configured in time, but reading RTC counter was too late. In such case the requested task would be triggered in HW, but function would behave like it was not.

Reply:
Thanks for pointing out. I pushed a fix for race condition